### PR TITLE
feat: gas price dialog

### PR DESCRIPTION
**Pull request by zhouhanseng**
image

image

image

### CODE REVIEW

File: packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    },
+}))
+
+export function GasNowDialog() {
+    const classes = useStyles()
+    const { t } = useI18N()
+    const gasNow = useValueRef(currentGasNowSettings)
+    const rapidGasFirstCache = useMemo(() => gasNow?.rapid, [])
+    const [customGas, setCustomGas] = useState(formatWeiToGwei(rapidGasFirstCache ?? 0).toString())
+    const customGasToWei = Number(customGas === '' ? 0 : customGas) * 10 ** 9
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx:95]: Please use a big number.

File: packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    const customEstimateTime = gasNow
+        ? customGasToWei >= gasNow.rapid
+            ? t('plugin_gas_now_dialog_sec', { time: 15 })
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx:98]: Does the time need to read dynamically?

File: packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+            : customGasToWei >= gasNow.fast
+            ? t('plugin_gas_now_dialog_min', { time: 1 })
+            : customGasToWei >= gasNow.standard
+            ? t('plugin_gas_now_dialog_min', { time: 3 })
+            : t('plugin_gas_now_dialog_min', { time: '>10' })
+        : ''
+    const { open, closeDialog, setDialog } = useRemoteControlledDialog(
+        EthereumMessages.events.gasPriceDialogUpdated,
+        (_ev) => void 0,
+    )
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx:108]:
```suggestion
    const { open, closeDialog, setDialog } = useRemoteControlledDialog(
        EthereumMessages.events.gasPriceDialogUpdated
    )
```

File: packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    const { value: detailedTokens } = useAssets([])
+    const [select, setSelect] = useState(gasNow ? 1 : 2)
+    const nativeToken = detailedTokens[0]
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx:111]: `useNativeTokenDetailed`

File: packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            : customGasToWei >= gasNow.standard
+            ? t('plugin_gas_now_dialog_min', { time: 3 })
+            : t('plugin_gas_now_dialog_min', { time: '>10' })
+        : ''
+    const { open, closeDialog, setDialog } = useRemoteControlledDialog(
+        EthereumMessages.events.gasPriceDialogUpdated,
+        (_ev) => void 0,
+    )
+    const { value: detailedTokens } = useAssets([])
+    const [select, setSelect] = useState(gasNow ? 1 : 2)
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx:110]: Refactor it by using an `enum`.

File: packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+                                    <Skeleton animation="wave" variant="rectangular" height={15} width="80%"></Skeleton>
+                                    <Skeleton animation="wave" variant="rectangular" height={15} width="80%"></Skeleton>
+                                    <Skeleton animation="wave" variant="rectangular" height={12} width="40%"></Skeleton>
+                                </>
+                            )}
+                        </ListItem>
+                    ))}
+                </List>
+                <Typography className={classes.intro}>{t('plugin_gas_now_dialog_intro')}</Typography>
+                <EthereumWalletConnectedBoundary>
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx:219]:
```suggestion
                <EthereumWalletConnectedBoundary offChain>
```

File: packages/maskbook/src/plugins/Wallet/settings.ts [hunk 7]
```diff
@@ -100,6 +98,41 @@ export const currentNonceSettings = createGlobalSettings<number>(`${PLUGIN_IDENT
...
+        console.log('GasNow websocket connected.')
+    }
+    gasNowSocket.onmessage = (event) => {
+        const gasNow: GasNow = JSON.parse(event.data).data.gasPrices
+        currentGasNowSettings.value = gasNow
+    }
+    gasNowSocket.onclose = () => {
+        console.log('GasNow websocket closed, try to reconnect...')
+        currentGasNowSettings.value = null
+        setTimeout(connectGasNow, 1000)
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Wallet/settings.ts:135]: Should this timer be cleared somewhere?

**Comment by guanbinrui** [comment]
Why the `pnpm-lock.yaml` file was changed?

File: packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+    },
+    dotSelector: {
+        color: '#1C68F3',
+    },
+    customInput: {
+        width: 60,
+        marginRight: theme.spacing(1),
+    },
+    button: {
+        margin: '16px auto',
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx:84]:
```suggestion
        margin: `${theme.spacing(2)} auto`,
```

File: packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+                                    className={classNames(classes.dot, i === select ? classes.dotSelector : undefined)}
+                                />
+                            </div>
+                            {gasNow || i === 2 ? (
+                                <>
+                                    <div>
+                                        {i == 2 ? (
+                                            <div className={classes.gweiBox}>
+                                                <TextField
+                                                    className={classes.customInput}
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx:177]: Do we need to set a max value for this field?

File: packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx (hunk 2 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx:98]: The time is a fixed value.

File: packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx (hunk 9 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx:177]: I don't think it is needed

**Comment by zhouhanseng** [comment]
> Why the `pnpm-lock.yaml` file was changed?

single quote become double quote '@commitlint/cli' => "@commitlint/cli", I am going to revert this.

File: packages/maskbook/src/plugins/Wallet/settings.ts (hunk 7 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Wallet/settings.ts:135]: I think ti can be clear at `onopen()`

File: packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx (hunk 4 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx:111]: `useNativeTokenDetailed` can not get the usd price.

File: packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx (hunk 5 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Ethereum/UI/GasNowDialog.tsx:107]: I will refactor this later

**Pull request merged**